File: rfwhisper/models/registry.py
"""
Model registry: artifact paths, execution-provider auto-select, ONNX session options.

`ARTIFACTS` and `model_path` describe on-disk locations for the fetch script (#12).
`resolve_providers`, `make_session_options`, and `load_model` wire those weights into
the realtime path with the correct EP priority for each OS.

Licensing: DeepFilterNet upstream is MIT or Apache-2.0; ONNX conversions inherit that.
See models/README.md and docs/datasets/SAMPLE_TRAINING_DATASET.md.
"""

from __future__ import annotations

import importlib
import platform
import warnings
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from rfwhisper.models.base import Model
from rfwhisper.models.null_model import NullModel


def _import_onnxruntime() -> Any:
    """
    Import onnxruntime via importlib so mypy doesn't choke on the missing stubs.

    onnxruntime ships no type information; the project's other call sites use
    `# type: ignore[import-untyped]`, but the right ignore-code differs between
    mypy 1.10 and 1.20. Routing through importlib avoids the mismatch.
    """
    return importlib.import_module("onnxruntime")


REPO_ROOT = Path(__file__).resolve().parents[2]


@dataclass(frozen=True)
class ModelArtifact:
    """
    On-disk descriptor for a downloadable model artifact.

    Consumed by [fetch.py](rfwhisper/models/fetch.py) (#12). `sha256` is the literal
    `"VERIFY_ON_FIRST_RUN"` until a release pins a hash — see `models/README.md` for
    the bless-on-first-run flow.
    """

    name: str
    relpath: str
    url: str
    sha256: str  # set to "VERIFY_ON_FIRST_RUN" until release pins a hash
    license_note: str


ARTIFACTS: tuple[ModelArtifact, ...] = (
    ModelArtifact(
        name="DeepFilterNet3 ONNX (community export — verify before production)",
        relpath="models/deepfilternet3/dfn3.onnx",
        url="https://huggingface.co/aufklarer/DeepFilterNet3-ONNX/resolve/main/deepfilter.onnx",
        sha256="VERIFY_ON_FIRST_RUN",
        license_note="Upstream DeepFilterNet: MIT OR Apache-2.0. Third-party export; test parity.",
    ),
    ModelArtifact(
        name="ERB + window auxiliary (if required by your ONNX pack)",
        relpath="models/deepfilternet3/deepfilter-auxiliary.bin",
        url="https://huggingface.co/aufklarer/DeepFilterNet3-ONNX/resolve/main/deepfilter-auxiliary.bin",
        sha256="VERIFY_ON_FIRST_RUN",
        license_note="See HuggingFace model card; must match the ONNX you run.",
    ),
)


def model_path(name: str) -> Path:
    """Return the absolute on-disk directory for a given model name (e.g. `deepfilternet3`)."""
    return REPO_ROOT / "models" / name


# --- Execution-provider auto-select ----------------------------------------------------

# Per-OS EP priority. CUDA / XNNPACK / CPU are appended on every OS so the resolver
# falls through cleanly when the preferred EP isn't installed.
_OS_PREFERRED: dict[str, tuple[str, ...]] = {
    "Darwin": ("CoreMLExecutionProvider",),
    "Windows": ("DmlExecutionProvider",),
    "Linux": (),
}
_COMMON_FALLBACK: tuple[str, ...] = (
    "CUDAExecutionProvider",
    "XnnpackExecutionProvider",
    "CPUExecutionProvider",
)


def resolve_providers(
    *,
    system: str | None = None,
    available: list[str] | None = None,
) -> list[str]:
    """
    Return ONNX Runtime execution-provider names in priority order.

    Picks the OS-preferred EP first (CoreML on macOS, DirectML on Windows), then falls
    back through CUDA → XNNPACK → CPU. EPs not present in `available` are skipped, so a
    machine without DirectML installed silently falls through to the next option rather
    than crashing.

    `system` and `available` are injected for tests; production calls leave them None.
    """
    sys_name = system if system is not None else platform.system()
    avail = available if available is not None else _query_available_providers()

    priority: list[str] = []
    priority.extend(_OS_PREFERRED.get(sys_name, ()))
    priority.extend(_COMMON_FALLBACK)

    selected = [p for p in priority if p in avail]
    if not selected:
        # No EPs matched — onnxruntime will always advertise CPU, so this is only
        # reachable when caller mocked an empty list. Force CPU so we never hand
        # InferenceSession an empty providers= list.
        warnings.warn(
            "No matching ONNX execution providers; defaulting to CPUExecutionProvider",
            RuntimeWarning,
            stacklevel=2,
        )
        return ["CPUExecutionProvider"]
    return selected


def _query_available_providers() -> list[str]:
    """
    Query installed ONNX Runtime EPs, or return `["CPUExecutionProvider"]` when ORT
    isn't installed at all (NullModel-only environments).
    """
    try:
        ort = _import_onnxruntime()
    except ImportError:
        return ["CPUExecutionProvider"]
    providers: list[str] = list(ort.get_available_providers())
    return providers


def make_session_options() -> Any:
    """
    Build SessionOptions tuned for realtime per AGENTS §Real-Time Constraints.

    Returns ort.SessionOptions; typed as Any so this module imports without
    onnxruntime present (NullModel path on lean CI runners).
    """
    ort = _import_onnxruntime()
    opts = ort.SessionOptions()
    opts.intra_op_num_threads = 2
    opts.inter_op_num_threads = 1
    opts.enable_cpu_mem_arena = True
    opts.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
    return opts


# --- Name → Model factory --------------------------------------------------------------

# Relative path (under REPO_ROOT) to the primary weights file for each named model.
# Presence on disk is what gates "have weights" vs "need fallback".
_WEIGHTS: dict[str, str] = {
    "deepfilternet3": "models/deepfilternet3/dfn3.onnx",
    "rnnoise": "models/rnnoise/model.onnx",
}

ModelFactory = Callable[[Path], Model]


def _load_dfn3(weights: Path) -> Model:
    """Construct the DFN3 wrapper (#10). Raises ImportError until that wrapper lands."""
    from rfwhisper.models.dfn3 import DFN3  # type: ignore[import-not-found]

    return DFN3(  # type: ignore[no-any-return]
        weights,
        providers=resolve_providers(),
        session_options=make_session_options(),
    )


def _load_rnnoise(weights: Path) -> Model:
    """Construct the RNNoise wrapper (#11). Raises ImportError until that wrapper lands."""
    from rfwhisper.models.rnnoise import RNNoise  # type: ignore[import-not-found]

    return RNNoise(  # type: ignore[no-any-return]
        weights,
        providers=resolve_providers(),
        session_options=make_session_options(),
    )


_FACTORIES: dict[str, ModelFactory] = {
    "deepfilternet3": _load_dfn3,
    "rnnoise": _load_rnnoise,
}


def load_model(name: str, *, fallback_to_null: bool = False) -> Model:
    """
    Resolve `name` to a Model instance.

    Names: `null`, `deepfilternet3` (#10), `rnnoise` (#11). When weights or the wrapper
    module aren't present, `fallback_to_null=True` returns NullModel (with a warning);
    otherwise FileNotFoundError or ImportError propagates.
    """
    if name == "null":
        return NullModel()

    if name not in _FACTORIES:
        raise ValueError(f"Unknown model: {name!r}. Known: {sorted(['null', *_FACTORIES])}")

    weights = REPO_ROOT / _WEIGHTS[name]
    if not weights.is_file():
        return _fallback_or_raise(
            name,
            FileNotFoundError(f"Weights for {name!r} not found at {weights}"),
            fallback_to_null,
        )

    try:
        return _FACTORIES[name](weights)
    except ImportError as exc:
        return _fallback_or_raise(name, exc, fallback_to_null)


def _fallback_or_raise(name: str, exc: Exception, fallback_to_null: bool) -> Model:
    """Return NullModel with a warning when fallback is allowed; otherwise re-raise `exc`."""
    if fallback_to_null:
        warnings.warn(
            f"load_model({name!r}) falling back to NullModel: {exc}",
            RuntimeWarning,
            stacklevel=3,
        )
        return NullModel()
    raise exc

File: src/models/registry.rs
//! Model registry: artifact paths, execution-provider auto-select, thread knobs.
//!
//! `ARTIFACTS` and `model_path` describe on-disk locations for the fetch command.
//! `resolve_providers` keeps the per-OS ONNX Runtime execution-provider priority
//! logic (CoreML on macOS, DirectML on Windows, CUDA → XNNPACK → CPU fallback) so
//! the DFN3 backend can wire straight into it when it lands.
//!
//! Licensing: DeepFilterNet upstream is MIT or Apache-2.0; ONNX conversions
//! inherit that. See models/README.md.

use std::path::{Path, PathBuf};

use crate::models::{Model, NullModel};

/// Default per AGENTS §Real-Time Constraints; override via `RFWHISPER_ORT_INTRA_OP`.
const DEFAULT_INTRA_OP_THREADS: u32 = 2;
const INTRA_OP_ENV_VAR: &str = "RFWHISPER_ORT_INTRA_OP";

#[derive(Debug, thiserror::Error)]
pub enum RegistryError {
    #[error("Unknown model: {0:?}. Known: [\"deepfilternet3\", \"null\", \"rnnoise\"]")]
    UnknownModel(String),
    #[error("Weights for {name:?} not found at {path}")]
    WeightsNotFound { name: String, path: PathBuf },
    #[error(
        "{0:?} backend is not yet available in the Rust port (tracked in the DFN3 / RNNoise \
         backend issues)"
    )]
    BackendUnavailable(String),
}

/// On-disk descriptor for a downloadable model artifact.
///
/// Consumed by [`crate::models::fetch`]. `sha256` is the literal
/// `"VERIFY_ON_FIRST_RUN"` until a release pins a hash — see `models/README.md`
/// for the bless-on-first-run flow.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ModelArtifact {
    pub name: &'static str,
    pub relpath: &'static str,
    pub url: &'static str,
    /// Set to `"VERIFY_ON_FIRST_RUN"` until release pins a hash.
    pub sha256: &'static str,
    pub license_note: &'static str,
}

pub const ARTIFACTS: [ModelArtifact; 2] = [
    ModelArtifact {
        name: "DeepFilterNet3 ONNX (community export — verify before production)",
        relpath: "models/deepfilternet3/dfn3.onnx",
        url: "https://huggingface.co/aufklarer/DeepFilterNet3-ONNX/resolve/main/deepfilter.onnx",
        sha256: "VERIFY_ON_FIRST_RUN",
        license_note: "Upstream DeepFilterNet: MIT OR Apache-2.0. Third-party export; test parity.",
    },
    ModelArtifact {
        name: "ERB + window auxiliary (if required by your ONNX pack)",
        relpath: "models/deepfilternet3/deepfilter-auxiliary.bin",
        url: "https://huggingface.co/aufklarer/DeepFilterNet3-ONNX/resolve/main/deepfilter-auxiliary.bin",
        sha256: "VERIFY_ON_FIRST_RUN",
        license_note: "See HuggingFace model card; must match the ONNX you run.",
    },
];

/// Repository root: `RFWHISPER_ROOT` when set, else the crate's source checkout.
///
/// The Python package derived this from `__file__`; the env override is the
/// escape hatch for installed binaries that keep their models elsewhere.
pub fn repo_root() -> PathBuf {
    if let Ok(root) = std::env::var("RFWHISPER_ROOT") {
        return PathBuf::from(root);
    }
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
}

/// Absolute on-disk directory for a given model name (e.g. `deepfilternet3`).
pub fn model_path(name: &str) -> PathBuf {
    repo_root().join("models").join(name)
}

// --- Execution-provider auto-select ---------------------------------------------------

/// Per-OS EP priority. CUDA / XNNPACK / CPU are appended on every OS so the resolver
/// falls through cleanly when the preferred EP isn't installed.
fn os_preferred(system: &str) -> &'static [&'static str] {
    match system {
        "Darwin" | "macos" => &["CoreMLExecutionProvider"],
        "Windows" | "windows" => &["DmlExecutionProvider"],
        _ => &[],
    }
}

const COMMON_FALLBACK: [&str; 3] = [
    "CUDAExecutionProvider",
    "XnnpackExecutionProvider",
    "CPUExecutionProvider",
];

fn current_system() -> &'static str {
    match std::env::consts::OS {
        "macos" => "Darwin",
        "windows" => "Windows",
        _ => "Linux",
    }
}

/// Return ONNX Runtime execution-provider names in priority order.
///
/// Picks the OS-preferred EP first (CoreML on macOS, DirectML on Windows), then
/// falls back through CUDA → XNNPACK → CPU. EPs not present in `available` are
/// skipped, so a machine without DirectML installed silently falls through to the
/// next option rather than crashing.
///
/// `system` and `available` are injected for tests; production calls pass `None`.
pub fn resolve_providers(system: Option<&str>, available: Option<&[&str]>) -> Vec<String> {
    let sys_name: &str = match system {
        Some(s) => s,
        None => current_system(),
    };
    let queried;
    let avail: &[&str] = match available {
        Some(a) => a,
        None => {
            queried = query_available_providers();
            &queried
        }
    };

    let mut priority: Vec<&str> = Vec::new();
    priority.extend_from_slice(os_preferred(sys_name));
    priority.extend_from_slice(&COMMON_FALLBACK);

    let selected: Vec<String> = priority
        .into_iter()
        .filter(|p| avail.contains(p))
        .map(str::to_string)
        .collect();
    if selected.is_empty() {
        // No EPs matched — ONNX Runtime always advertises CPU, so this is only
        // reachable when a caller injected an empty list. Force CPU so we never
        // hand an inference session an empty provider list.
        eprintln!(
            "warning: no matching ONNX execution providers; defaulting to CPUExecutionProvider"
        );
        return vec!["CPUExecutionProvider".to_string()];
    }
    selected
}

/// Installed ONNX Runtime EPs. Until the ONNX backend lands in the Rust port this
/// is always the CPU provider (the NullModel-only environment).
fn query_available_providers() -> Vec<&'static str> {
    vec!["CPUExecutionProvider"]
}

/// Resolve the intra-op thread count, honoring `RFWHISPER_ORT_INTRA_OP` if set.
///
/// Invalid values warn and fall back to the default rather than crashing the engine.
pub fn resolve_intra_op_threads() -> u32 {
    let raw = match std::env::var(INTRA_OP_ENV_VAR) {
        Ok(v) => v,
        Err(_) => return DEFAULT_INTRA_OP_THREADS,
    };
    match raw.parse::<i64>() {
        Ok(n) if n >= 1 => n as u32,
        Ok(n) => {
            eprintln!(
                "warning: {INTRA_OP_ENV_VAR}={n} must be >= 1; using default \
                 {DEFAULT_INTRA_OP_THREADS}"
            );
            DEFAULT_INTRA_OP_THREADS
        }
        Err(_) => {
            eprintln!(
                "warning: {INTRA_OP_ENV_VAR}={raw:?} is not an int; using default \
                 {DEFAULT_INTRA_OP_THREADS}"
            );
            DEFAULT_INTRA_OP_THREADS
        }
    }
}

// --- Name → Model factory --------------------------------------------------------------

/// Relative path (under the repo root) to the primary weights file for each named
/// model. Presence on disk is what gates "have weights" vs "need fallback".
fn weights_relpath(name: &str) -> Option<&'static str> {
    match name {
        "deepfilternet3" => Some("models/deepfilternet3/dfn3.onnx"),
        "rnnoise" => Some("models/rnnoise/model.onnx"),
        _ => None,
    }
}

/// Resolve `name` to a `Model` instance using the default repo root.
///
/// Names: `null`, `deepfilternet3`, `rnnoise`. When weights or the backend aren't
/// present, `fallback_to_null = true` returns `NullModel` (with a warning);
/// otherwise the error propagates.
pub fn load_model(name: &str, fallback_to_null: bool) -> Result<Box<dyn Model>, RegistryError> {
    load_model_in(&repo_root(), name, fallback_to_null)
}

/// [`load_model`] with an explicit root — injected by tests, mirrored from the
/// Python tests' `REPO_ROOT` monkeypatching.
pub fn load_model_in(
    root: &Path,
    name: &str,
    fallback_to_null: bool,
) -> Result<Box<dyn Model>, RegistryError> {
    if name == "null" {
        return Ok(Box::new(NullModel));
    }
    let Some(rel) = weights_relpath(name) else {
        return Err(RegistryError::UnknownModel(name.to_string()));
    };

    let weights = root.join(rel);
    if !weights.is_file() {
        return fallback_or_raise(
            name,
            RegistryError::WeightsNotFound {
                name: name.to_string(),
                path: weights,
            },
            fallback_to_null,
        );
    }

    // Weights exist but the NN backends have not landed in the Rust port yet —
    // same shape as the Python ImportError path for a missing wrapper module.
    fallback_or_raise(
        name,
        RegistryError::BackendUnavailable(name.to_string()),
        fallback_to_null,
    )
}

/// Return `NullModel` with a warning when fallback is allowed; otherwise the error.
fn fallback_or_raise(
    name: &str,
    err: RegistryError,
    fallback_to_null: bool,
) -> Result<Box<dyn Model>, RegistryError> {
    if fallback_to_null {
        eprintln!("warning: load_model({name:?}) falling back to NullModel: {err}");
        return Ok(Box::new(NullModel));
    }
    Err(err)
}
